packages/assert/src/assert.js

```diff
@@ -0,0 +1,378 @@
+  const seenSet = new Set();
+  const replacer = (_, val) => {
+    if (typeof val === 'object' && val !== null) {
+      if (seenSet.has(val)) {
+        return '<**seen**>';
+      }
+      seenSet.add(val);
+    }
+    return val;
+  };
+  return JSON.stringify(payload, replacer);
+};
+freeze(cycleTolerantStringify);
+export { cycleTolerantStringify };
+
+const declassifiers = new WeakMap();
+
+/**
+ * To "declassify" and quote a substitution value used in a
+ * details`...` template literal, enclose that substitution expression
+ * in a call to `q`. This states that the argument should appear quoted (with
+ * `JSON.stringify`), in the error message of the thrown error. The payload
+ * itself is still passed unquoted to the console as it would be without q.
+ *
+ * Starting from the example in the `details` comment, say instead that the
+ * color the sky is supposed to be is also computed. Say that we still don't
+ * want to reveal the sky's actual color, but we do want the thrown error's
+ * message to reveal what color the sky was supposed to be:
+ * ```js
+ * assert.equal(
+ *   sky.color,
+ *   color,
+ *   details`${sky.color} should be ${q(color)}`,
+ * );
+ * ```
+ *
+ * @typedef {Object} StringablePayload
+ * @property {() => string} toString How to print the payload
+ *
+ * @param {*} payload What to declassify
+ * @returns {StringablePayload} The declassified payload
+ */
+const q = payload => {
+  // Don't harden the payload
+  const result = freeze({
+    toString: freeze(() => cycleTolerantStringify(payload)),
+  });
+  declassifiers.set(result, payload);
+  return result;
+};
+freeze(q);
+export { q };
+
+// /////////////////////////////////////////////////////////////////////////////
+
+/**
+ * @enum {ErrorInfoKind}
+ */
+const ErrorInfo = {
+  // @ts-ignore TODO Why doesn't this type check?
+  NOTE: 'ERROR_NOTE:',
+  // @ts-ignore TODO Why doesn't this type check?
+  MESSAGE: 'ERROR_MESSAGE:',
+};
+freeze(ErrorInfo);
+export { ErrorInfo };
+
+/**
+ * Tell the `console` to consider the `error` to have an annotation of
+ * `errorInfoKind` whose contents are the logArgs returned by calling
+ * `getLogArgs()`.
+ *
+ * This function feature detects whether the console has (non-standard) a
+ * `rememberErrorInfo`. The consoles made by the `@agoric/console` package
+ * have this custom method.
+ * This function then calls this custom method to inform the console. For a
+ * normal console without a `rememberErrorInfo` method, this extra error info
+ * is ignored.
+ */
+const logErrorInfoToConsole = (console, error, errorInfoKind, getLogArgs) => {
+  /**
+   * @type {RememberErrorInfo}
+   */
+  const rememberErrorInfo = console.rememberErrorInfo;
+  if (typeof rememberErrorInfo === 'function') {
+    // A console with a custom `rememberErrorInfo` method can avoid even
+    // calling `getLogArgs()` until it needs to, which most won't.
+    rememberErrorInfo(error, errorInfoKind, getLogArgs);
+  }
+};
```

Comment:
It looks like `errorInfoKind` could be folded into the return of `getLogArgs` to simplify the signature of `rememberErrorInfo`.

Since `assert.note` is driving, it might be good for the console method to have a related name like `console.attachNote`, `deferNote`, or even just `note`.

Design note: I gather that the log array callback allows the `console` to return control to the `assert` module to expand details.

Maybe this could be more direct, like a `console.note(err, callback())`, where the only behavior is that the console will call the assertion library back if it encounters the given error object. Then, in the callback, `assert` would use the `console` API directly. Perhaps that would produce less specific coupling between `console` and `assert`.

Reply:
I didn't follow this. Let's discuss.

Reply:
> It looks like `errorInfoKind` could be folded into the return of `getLogArgs` to simplify the signature of `rememberErrorInfo`.
>
> Since `assert.note` is driving, it might be good for the console method to have a related name like `console.attachNote`, `deferNote`, or even just `note`.

`rememberErrorInfo` is gone

> Design note: I gather that the log array callback allows the `console` to return control to the `assert` module to expand details.

Callbacks work differently. PTAL

> Maybe this could be more direct, like a `console.note(err, callback())`, where the only behavior is that the console will call the assertion library back if it encounters the given error object. Then, in the callback, `assert` would use the `console` API directly. Perhaps that would produce less specific coupling between `console` and `assert`.

I did parameterize `makeCausalConsole` with a `loggedErrorHandler`, a privileged export from the `assert` package, so that neither `console` nor `assert` have any "extra" API surface. However, I did it differently than we talked about. I left the causal console in control of the chaining of error reports. The causal console primarily calls back into the `loggedErrorHandler` to access to `assert` module's hidden bookkeeping.

Reply:
Not resolving yet due to the PTAL.